audit/audit_contract.go

```diff
@@ -36,39 +36,50 @@ func (a auditHeader) ID() []byte {
 	return append(buf.([]byte), append(a.cid, shortedKey...)...)
 }
 
-const (
-	netmapContractKey = "netmapScriptHash"
-
-	notaryDisabledKey = "notary"
-)
-
 // nolint:deadcode,unused
 func _deploy(data interface{}, isUpdate bool) {
 	ctx := storage.GetContext()
 	if isUpdate {
 		args := data.([]interface{})
 		common.CheckVersion(args[len(args)-1].(int))
+
+		switchToNotary(ctx, args)
+
 		return
 	}
 
 	args := data.(struct {
 		notaryDisabled bool
-		addrNetmap     interop.Hash160
+		_              interop.Hash160 // Netmap contract from non-notary legacy
 	})
 
-	if len(args.addrNetmap) != interop.Hash160Len {
-		panic("incorrect length of contract script hash")
+	if args.notaryDisabled {
+		panic("non-notary mode is not supported anymore")
 	}
 
-	storage.Put(ctx, netmapContractKey, args.addrNetmap)
+	runtime.Log("audit contract initialized")
+}
 
-	// initialize the way to collect signatures
-	storage.Put(ctx, notaryDisabledKey, args.notaryDisabled)
-	if args.notaryDisabled {
-		runtime.Log("audit contract notary disabled")
+// re-initializes contract from non-notary to notary mode. Does nothing if
+// action has already been done. The function is called on contract update with
+// storage.Context and parameters from _deploy.
+//
+// switchToNotary removes values stored by 'netmapScriptHash' and 'notary' keys.
+func switchToNotary(ctx storage.Context, args []interface{}) {
+	const notaryDisabledKey = "notary" // non-notary legacy
+
+	notaryVal := storage.Get(ctx, notaryDisabledKey)
+	if notaryVal == nil {
+		runtime.Log("contract is already notarized")
+		return
 	}
 
-	runtime.Log("audit contract initialized")
+	storage.Delete(ctx, notaryDisabledKey)
+	storage.Delete(ctx, "netmapScriptHash")
```

Comment:
Safe to do on notary-enabled networks?

Reply:
I guess, these are only usages:
https://github.com/nspcc-dev/neofs-contract/blob/5b007f429fa980ccd3879e328010aa771692e320/audit/audit_contract.go#L98-L100
https://github.com/nspcc-dev/neofs-contract/blob/5b007f429fa980ccd3879e328010aa771692e320/audit/audit_contract.go#L181-L202